Topbar redux

goldylucks opened the pull request:
### NOTES:
1.  implement reudx
2. map state / dispatch to props in components*
3. reuse from components: Account, Identicon, WebCompositor instead of Web3Component
4. naming changes
   - [allAcounts] -> [accounts]
   - [accounts] -> activeAccount
   - isSignerEnabled -> signerPort
   - web3 -> exposedWeb3
   - rawWeb3 -> internalWeb3
   - DappNav -> DappsNav
5. (components)
   - add util/logger
   - add AccountWeb3
   - add click support to Toast
   - add toastr, logger -> actions, reducer, middleware
   - add util/util funcs -> toPromise, identity, stringifyIfObject
6. (extension)
   - response 'version' instead of { version: 'version' } to website calls
7. Apply conventions: React.Component, React.PropTypes -> Component, PropTypes
### TODO:
- [x] fix css loading of components library
- [x] implement pendingTransaction (intercepting when signer disabled)
- [x] test toasts disappearing automatically

*
This is a somewhat different approach than what we did so far.
for example in createAccount component:
Parent - Topbar
Action - create account

Topbar state isn't related to create account, so instead of using connect to map the action in Topbar and pass it to createAccount, the action is mapped directly in createAccount

I've seen this approach in a few articles and it makes sense (on sight)

**
after component dispatches the createAccount action, it needs to reflect the response in the UI. might b possible with redux-thunk

goldylucks committed:
(components): add toastr and logger actions and middlewards

goldylucks committed:
(home): add redux infrastructure

goldylucks committed:
(Topbar): implement redux

goldylucks committed:
Merge branch 'master' into topbar-redux

# Conflicts:
#	home/web/client/components/Root/Root.js
#	home/web/client/components/Toasts/Toasts.js
#	home/web/webpack.config.js

goldylucks committed:
(TopBar): move fixAccountNames from utils to storage middleware

goldylucks committed:
(home): parse components css

tomusdrw commented on status/web/client/provider/web3-provider.js:
```diff
@@ -132,12 +132,12 @@ export class Web3Provider extends Web3Base {
   }
 
-  actionProp (action) {
+  actionToStatePtop (action) {
```
typo

tomusdrw commented on status/web/client/middleware/localstorage.js:
```diff
@@ -2,7 +2,10 @@
 import { syncRpcStateFromLocalStorage } from '../actions/localstorage';
 import rpcMetods from '../data/rpc.json';
-
+sdf 
```
Probably doesn't work.

goldylucks committed:
cleanup

goldylucks committed:
(TopBar) fix toasts and transcation confirmation

goldylucks committed:
(TopBar): revert Account to not import from components

goldylucks committed:
Merge branch 'master' into topbar-redux

# Conflicts:
#	home/web/package.json

goldylucks committed:
(Topbar): fix fetch extension version

goldylucks committed:
(TopBar): revert to local Account (not from components)

goldylucks committed:
Merge branch 'master' into topbar-redux

# Conflicts:
#	home/web/package.json

goldylucks committed:
(TopBar): compiled js files

tomusdrw commented on components/src/middlewares/Logger.js:
```diff
@@ -0,0 +1,25 @@
+    const seconds = date.getSeconds();
+    const minutes = date.getMinutes();
+    const hour = date.getHours();
+    return `${hour}::${minutes}::${seconds}`;
```
Leading zeros will be missing.

goldylucks replied:
added pad function

```
  now () {
    const date = new Date(Date.now());
    const seconds = this.pad(date.getSeconds());
    const minutes = this.pad(date.getMinutes());
    const hour = this.pad(date.getHours());
    return `${hour}::${minutes}::${seconds}`;
  }

  pad (n) {
    return n < 10 ? '0' + n : n;
  }

```

tomusdrw commented on components/src/middlewares/Toastr.spec.js:
```diff
@@ -0,0 +1,174 @@
+
+      // then
+      expect(cut._timeouts[String(toastNo)]).to.be.an('object');
+      setTimeout(() => {
```
http://sinonjs.org/docs/#clock

goldylucks replied:
we have this in 3 more places in our codebase
let's make it in a separate PR

tomusdrw commented on components/src/middlewares/Toastr.spec.js:
```diff
@@ -0,0 +1,174 @@
+    });
+  });
+
+  describe('SET TIMEOUT FOR', () => {
```
Those tests are relying on too much private stuff for me. Would be better to test a general public API then to verify that some particular values are present in internal structures.
(but it's just an opinion, the tests are ok)

tomusdrw commented on components/src/util/logger.js:
```diff
@@ -0,0 +1,22 @@
+}
+
+function devLogger () {
+  return console;
```
Would be good to keep the same methods for prod and dev:

``` javascript
return {
  log: console.log.bind(console),
  info: ..,
  error: ...,
  warn: ...
};
```

goldylucks replied:
what do u mean same methods?

tomusdrw replied:
So right now I can do `logger.debug()` <- which will work in development but will be broken in production. Production and Development loggers should have the same APIs, otherwise you're production might be not working, but dev will.

goldylucks replied:
kk
let's have it in a separate PR, this logger is just reused from previous projects

tomusdrw commented on components/src/util/util.js:
```diff
@@ -0,0 +1,23 @@
+  });
+}
+
+export function stringifyIfObject (any) {
```
Not used anywhere?

goldylucks replied:
in status page

tomusdrw replied:
How come? It's a new file?

goldylucks replied:
yes it was migrated
removed it now anyway

will remove it form status page as well, it's not really needed

tomusdrw commented on components/src/util/util.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+export function stringifyIfObject (any) {
+  if (typeof any === 'object') {
```
Most probably a bad method to detect objects. Have a look at `lodash.isObject`

goldylucks replied:
removed
will also remove from status in separate PR

goldylucks commented on home/web/client/reducers/pendingTransaction.js:
```diff
@@ -0,0 +1,48 @@
+    };
+  },
+
+  'reject transaction' (state, action) {
```
maybe change to `abort transcation`,
to distinguish from rejecting in signer

tomusdrw commented on home/web/client/actions/dapps.js:
```diff
@@ -0,0 +1,5 @@
+import { createAction } from 'redux-actions';
+
+export const error = createAction('error dapps');
```
What this action means?

goldylucks replied:
following our conventions:
'action context'

this means that there was an error with dapps, like we have in status page:
`export const error = createAction('error rpc')`

tomusdrw replied:
yeah, but I think it's part of fetching dapps, right?

We should keep it consistent for async actions for instance:

```
fetchDappsStarted
fetchDappsSuccess
fetchDappsErrors
```

or

```
startFetchingDapps
successFetchingDapps
errorFetchingDapps
```

goldylucks replied:
:+1: 
so just in rpc we have a special case with one error handler and no `start` action?

goldylucks commented on home/web/client/reducers/pendingTransaction.js:
```diff
@@ -0,0 +1,48 @@
+
+export default handleActions({
+
+  'update pendingTransaction' (state, action) {
```
Maybe change to `update unsignedTransaction`
to be consisted with rest of methods (and more accurate?)

tomusdrw commented on home/web/client/components/AccountsChooser/AccountsChooser.js:
```diff
@@ -1,89 +1,78 @@
-  state = {
-    defaultAccountIdx: 0
+  static contextTypes = {
+    web3: PropTypes.object.isRequired
```
Why does it still use `web3`?

goldylucks replied:
right, it is it's child Account which needs it
removed

tomusdrw commented on home/web/client/components/AccountsDetails/AccountsDetails.js:
```diff
@@ -1,40 +1,55 @@
+export default class AccountsDetails extends Component {
 
-  state = {};
+  static contextTypes = {
```
Not needed?

goldylucks replied:
line 77
`const address = this.context.web3.toChecksumAddress(acc)`

will b removed in future PR

goldylucks commented on home/web/client/reducers/pendingTransaction.js:
```diff
@@ -0,0 +1,48 @@
+import { handleActions } from 'redux-actions';
```
Change name to unsignedTransaction?

tomusdrw replied:
We can do it in separate PR, I'm ok with the naming now.

tomusdrw commented on home/web/client/components/CreateAccount/CreateAccount.js:
```diff
@@ -149,26 +169,45 @@ export default class CreateAccount extends Web3Component {
         primary
-        disabled={ !this.state.isValid }
-        onTouchTap={ this.submit }
+        disabled={ !this.state.isValid || this.state.isCreatingAccount }
```
Removing `isValid` from here will solve #63

goldylucks replied:
I thought about doing that in a separate PR with a UI fix
and removal of validations

tomusdrw replied:
Sure. Validations are ok though, they just shouldn't disable the button.

goldylucks commented on home/web/client/reducers/rpc.js:
```diff
@@ -0,0 +1,157 @@
+  highestBlock: 1234,
+  network: 'homestead',
+  createdAccount: '',
+  createdAccountError: ''
```
TBD:
separate `createdAccount` to own reducer

tomusdrw commented on home/web/client/components/Toasts/Toasts.js:
```diff
@@ -33,13 +32,41 @@ export default class Toasts extends Component {
   }
 
+  onClickToast = id => {
+    console.log('onClickToast', id);
```
REmove.

goldylucks replied:
done

tomusdrw commented on home/web/client/components/Toasts/Toasts.js:
```diff
@@ -33,13 +32,41 @@ export default class Toasts extends Component {
+  onClickToast = id => {
+    console.log('onClickToast', id);
+    this.props.onClickToast();
+    this.props.onRemoveToast(id);
```
Why is clicking the toast removing it?

goldylucks replied:
it opens the signer in a new tab and focuses on it
u c any reason to keep the toast open?
(it will b closed after 4 seconds anywayz)

tomusdrw replied:
:+1:

tomusdrw commented on home/web/client/components/Toasts/Toasts.js:
```diff
@@ -33,13 +32,41 @@ export default class Toasts extends Component {
 
+  onClickToast = id => {
+    console.log('onClickToast', id);
+    this.props.onClickToast();
```
Should have `id` parameter?

goldylucks replied:
nope
just opens signer

tomusdrw commented on home/web/client/components/TopBar/TopBar.js:
```diff
@@ -137,26 +100,28 @@ export default class TopBar extends Component {
+  onOpenDappsNav = () => {
     this.setState({
-      searchActive: active
+      DappsNavOpen: true
```
lowercase

goldylucks replied:
done

tomusdrw commented on home/web/client/components/TransactionConfirmation/TransactionConfirmation.js:
```diff
@@ -1,27 +1,34 @@
-    sending: false,
-    password: '',
-    error: false
+  static contextTypes = {
```
Not needed?

goldylucks replied:
needed for `this.context.web3.fromWei(value)` (removed by mistake)

tomusdrw commented on home/web/client/components/TransactionConfirmation/TransactionConfirmation.js:
```diff
@@ -97,101 +102,102 @@ export default class TransactionConfirmation extends Web3Component {
         hintText={ `Password for ${from}` }
         errorText={ errorMsg }
-        floatingLabelText='Unlock the account'
+        floatingLabelText='Type your password'
```
It's obvious that you need to type your password. It also says so in hint. We can use label to provide a context / give meaning to `type your password` action.

goldylucks replied:
is it accurate that it's unlocking the account?

what about in signer, should we put there something like `sign transaction` / `unlock the account` ?

tomusdrw replied:
I think `Type your password` is even more inaccurate - it will be displayed only after user already typed the password (cause it's floating text label) so it doesn't make sense.
It could be `Password to unlock your account`, I don't know what will work best.

goldylucks replied:
> `Password to unlock your account`

:+1:

tomusdrw commented on home/web/client/components/TransactionConfirmation/TransactionConfirmation.js:
```diff
@@ -97,101 +102,102 @@ export default class TransactionConfirmation extends Web3Component {
-  renderValue (value) {
-    const val = this.context.web3.fromWei(value);
+  renderValue () {
+    const { value } = this.props.transaction.params[0];
```
Will it be displayed correctly? It could do some conversion from wei to eth (and nice formatting too).

goldylucks replied:
fixed

tomusdrw commented on home/web/client/inject.js:
```diff
@@ -16,34 +15,78 @@ import web3extensions from 'dapps-react-components/src/util/web3.extensions';
-  </div>,
+  <Provider store={ store }>
+    <div className={ stylesReset.reset }>
+      <Web3Provider web3={ internalWeb3 }>
```
Is `Web3Provider` still needed?
Ah, ok for components from `dapps-react-components`

tomusdrw commented on home/web/client/inject.js:
```diff
@@ -16,34 +15,78 @@ import web3extensions from 'dapps-react-components/src/util/web3.extensions';
+const extensionProvider = new ExtensionProvider(store);
+extensionProvider.init();
+
+// IMPORTANT! Interceptor provider must be set in synchronous order after exposing web3
```
Might be better to actually move this before Rendering the components? And maybe `web3` should be exposed as close to this as possible?

goldylucks replied:
would it matter? it's all synchronous

tomusdrw replied:
But all stuff that this comment relates to will be close to each other. Also IMHO it makes more sense to first initialize store and middlewares and then render the state (instead of rendering possible uninitialized state, just to re-render it again right after).

goldylucks replied:
I followed the structure of status page:
store -> render dom -> providers
tho it makes sense to fire providers first.
u prefer all providers to init before rendering?

would b better to do that on status page as well (separate PR tho)

goldylucks replied:
I followed the structure of status page:
store -> render dom -> providers
tho it makes sense to fire providers first.
u prefer all providers to init before rendering?

would b better to do that on status page as well (separate PR tho)

tomusdrw commented on home/web/client/middlewares/rpc.js:
```diff
@@ -0,0 +1,92 @@
+        const { txRequest, password } = action.payload;
+        // swallow async action, dispatch sequence from middleware method
+        this.onStartSignTransaction(store, txRequest, password);
+        next(action);
```
The comment says that this should be swallowed?

goldylucks replied:
right, it was like that befre. I've changed the code a bit.
Removed.

tomusdrw commented on home/web/client/providers/interceptor.js:
```diff
@@ -0,0 +1,88 @@
+    };
+  }
+
+  stop () {
```
Not needed - middlewares don't have a lifecycle.

goldylucks replied:
removed

tomusdrw commented on home/web/client/providers/rpc.js:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  init () {
+    this.refreshTick();
```
It might be a good idea to do first `eth_accounts` call synchronously (but it's just a suggestion for future PR)

tomusdrw committed:
Merge branch 'topbar-redux' of github.com:ethcore/parity-ui into topbar-redux

tomusdrw committed:
Merge branch 'master' into topbar-redux

goldylucks committed:
(components): add react-hot-loader

its breaking without it when consuming a component from another project. adding webpack resolve didn't solve it.

goldylucks committed:
(Home): add eslint-config-standard-react

don't know when this was removed, or how it worked 'till now

goldylucks committed:
(TopBar): fixing #61 issues from @tomusdrw review

goldylucks committed:
Merge branch 'topbar-redux' of github.com:ethcore/parity-ui into topbar-redux

goldylucks committed:
(TopBar): fixing #61 issues from @tomusdrw 2th review

goldylucks committed:
(TopBar): fixing #61 issues from @tomusdrw 3th review

syncing accountsNames from topbar to wallet requires refresh, issue opened

goldylucks committed:
Merge branch 'master' into topbar-redux

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

goldylucks committed:
(TopBar): pushing #61 compiled files

goldylucks committed:
Merge branch 'master' into topbar-redux

# Conflicts:
#	extension/chrome/extension/Transactions/Transactions.js
#	signer/src/web/app.js
#	status/src/web/index.js

goldylucks committed:
(TopBar): #61 compiled files

goldylucks committed:
(TopBar): fix 61# failing tests
